modules/packages/MPI.chpl

```diff
@@ -29,7 +29,91 @@ routines ::
       MPI_Keyval_create
       MPI_Keyval_free
 
-since all of these are built around user-defined handlers, that we support.
+since all of these are built around user-defined handlers, that we do not support.
+
+We also include a few functions that are in later MPI standards. These are ::
+
+      MPI_Init_thread (and related constants)
+      MPI_Ibarrier
+
+Compiler wrappers
+-----------------
+
+Using MPI requires pointing the compiler to the location of the
+MPI headers, and including the appropriate MPI libraries. This is
+often done by using compiler wrappers like "mpicc". Setting
+``CHPL_TARGET_COMPILER=mpi-gnu`` and ``CHPL_TARGET_ARCH=none`` will
+get Chapel to use these compilers.
+
+Note that on Cray systems, this is not necessary, since the default
+compilers include the necessary MPI information.
+
+.. note ::
+  #. Currently, we only support the gnu compilers wrapped by MPI. We expect
+     to broaden this class in the future.
+  #. The current scripts fail to correctly determine the architecture for
+     the mpi-gnu suite, hence ``CHPL_TARGET_ARCH=none``. We expect this to be
+     fixed in the future.
+
+Interoperability Modes
+----------------------
+
+We envision two modes in which Chapel and MPI can interoperate with
+one another. The first of these is Chapel (local)+MPI. This has no
+constraints on the setup of Chapel and MPI, and is conceptually the same
+as OpenMP+MPI programming. All communication here is handled by
+MPI; Chapel is just used to provide local parallelism. Furthermore,
+in this mode, Chapel locales do not map to MPI ranks at all; the program
+just sees a single locale.
+Note that the same caveats that apply to OpenMP+MPI
+programming apply here. For instance, MPI must be initialized with the
+appropriate thread support (see below).
+
+The second mode of operation is where both Chapel and MPI communications
+may be mixed. This allows the user to use optimized/convenient MPI routines
+when desired, while having access to the Chapel PGAS model. Using this
+model requires
+
+1. The MPI library should (and in many cases must) support the MPI_THREAD_MULTIPLE
+   support level for threads. This is needed if the user plans on making
+   MPI calls from multiple threads. It is necessary when using the mpi-conduit
+   for GASNet, since the GASNet library may make simultaneous MPI calls.
+   Currently, the MPICH library (and its derivatives) appear to have the
+   most robust support for this.
```

Comment:
I think this statement is stronger than what we mean to say, since we really only tried OpenMPI and MPICH. I'd say "We've had good experience with MPI_THREAD_MULTIPLE support in MPICH".